### ISSUE TITLE
Scrollable Blocks should have gravity.

### ISSUE THREAD

**Issue by FlutterIssues** [flutter/flutter#136]
<a href="https://github.com/apwilson"><img src="https://avatars.githubusercontent.com/u/142874?v=3" align="left" width="96" height="96" hspace="10"></img></a> **Issue by [apwilson](https://github.com/apwilson)**
_Friday Aug 21, 2015 at 22:18 GMT_
_Originally opened as https://github.com/flutter/engine/issues/757_

----

**Comment by FlutterIssues** [comment]
<a href="https://github.com/Hixie"><img src="https://avatars.githubusercontent.com/u/551196?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Hixie](https://github.com/Hixie)**
_Saturday Aug 22, 2015 at 17:17 GMT_

----

I think the right API here is to just change the "direction" argument to take four values instead of two. Instead of horizontal/vertical, have it be right (what we do now for horizontal)/down (what we do now for vertical)/left (horizontal with reverse gravity)/up (vertical with reverse gravity).

**Comment by FlutterIssues** [comment]
<a href="https://github.com/abarth"><img src="https://avatars.githubusercontent.com/u/112007?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [abarth](https://github.com/abarth)**
_Saturday Aug 22, 2015 at 17:46 GMT_

----

Would you change the semantics of scrollOffset also?  For example, in the `up` direction, would offset zero correspond to the last item in the scrollable list being visible?  For example, [A, B, C, D, E, F, G], direction: up, scrollOffset: 0

```
-
C
D
E
F
G
-
```

Presumably we shouldn't reverse the order of the list, so [A, B, C], direction: up would display as:

```
-


A
B
C
-
```

Rather than:

```
-


C
B
A
-
```

**Comment by FlutterIssues** [comment]
<a href="https://github.com/Hixie"><img src="https://avatars.githubusercontent.com/u/551196?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Hixie](https://github.com/Hixie)**
_Saturday Aug 22, 2015 at 17:49 GMT_

----

I imagined that [A, B, C, D, E, F, G], direction: up, scrollOffset: 0 would render as:

```
-
E
D
C
B
A
-
```

**Comment by FlutterIssues** [comment]
<a href="https://github.com/abarth"><img src="https://avatars.githubusercontent.com/u/112007?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [abarth](https://github.com/abarth)**
_Saturday Aug 22, 2015 at 18:01 GMT_

----

That makes a certain amount of sense, but it would not solve @apwilson 's problem.  Maybe that means we should solve his problem in a different way.  Here's the effect that @apwilson wants:

Start out with a list [A, B, C, D, E, F, G, H, I, J]

```
-
D
E
F
G
H
-
```

Now, he wants to remove E from the list and end up with:

```
-
C
D
F
G
H
-
```

Now, he wants to remove D from the list and end up with:

```
-
C
F
G
H
I
-
```

In other words, he doesn't want the order of the list to change visually, he just wants to control whether new items are pulled into the viewport from the top or the bottom when the content size decreases.

**Comment by FlutterIssues** [comment]
<a href="https://github.com/Hixie"><img src="https://avatars.githubusercontent.com/u/551196?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Hixie](https://github.com/Hixie)**
_Saturday Aug 22, 2015 at 18:49 GMT_

----

I'm assuming based on our IRC conversation that when removing E and removing D, the gravities are configured differently.

What should happen when the list has only 5 items and one is removed in the reverse-gravity case?

**Comment by FlutterIssues** [comment]
<a href="https://github.com/abarth"><img src="https://avatars.githubusercontent.com/u/112007?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [abarth](https://github.com/abarth)**
_Saturday Aug 22, 2015 at 20:20 GMT_

----

I would expect empty space to appear from the top.

```
-

W
X
Y
Z
-
```

**Comment by FlutterIssues** [comment]
<a href="https://github.com/Hixie"><img src="https://avatars.githubusercontent.com/u/551196?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [Hixie](https://github.com/Hixie)**
_Saturday Aug 22, 2015 at 20:56 GMT_

----

That seems equivalent to reversing the list and using "direction:up", then, no? Or is the requirement really that different removals back to back cause the list to collapse in different directions?

**Comment by FlutterIssues** [comment]
<a href="https://github.com/abarth"><img src="https://avatars.githubusercontent.com/u/112007?v=3" align="left" width="48" height="48" hspace="10"></img></a> **Comment by [abarth](https://github.com/abarth)**
_Saturday Aug 22, 2015 at 21:17 GMT_

----

> That seems equivalent to reversing the list and using "direction:up", then, no?

Yes.

> Or is the requirement really that different removals back to back cause the list to collapse in different directions?

The same list needs to exhibit both collapsing directions at different times (but not at the same time).  It might be easier to show you what @apwilson is building than to work this all out via bugthread.

**Comment by collinjackson** [comment]
We need this to implement the "bottom compose" pattern

**Comment by abarth** [comment]
Ok.  I've hidden from this bug for long enough.

**Comment by abarth** [comment]
Warming up in https://github.com/flutter/flutter/pull/1710
